content/blog/2016-07-23-pull-does-not-scale-or-does-it.md

```diff
@@ -0,0 +1,175 @@
+discussion about monitoring systems and Prometheus's pull-based metrics
+collection approach comes up, someone inevitably chimes in about how a
+pull-based approach just "fundamentally doesn't scale". When asked why they
+think this is the case, they either respond with vague fears or cite reasons
+which only apply to systems which are fundamentally different from Prometheus
+in other ways. In fact, having worked with pull-based monitoring at the largest
+scales, this claim runs counter to our own operational experience.
+
+We already have an FAQ entry about
+[why Prometheus chooses pull over push](/docs/introduction/faq/#why-do-you-pull-rather-than-push?),
+but it does not focus specifically on scaling aspects. Let's have a closer look
+at the usual misconceptions around this claim and analyze whether and how they
+would apply to Prometheus.
+
+## Prometheus is not Nagios
+
+When people think of a monitoring system that actively pulls, they often think
+of Nagios. Nagios has a reputation of not scaling well, in part due to spawning
+subprocesses for active checks that can run arbitrary actions on the Nagios
+host in order to determine the health of a certain host or service. This sort
+of check architecture indeed does not scale well, as the central Nagios host
+quickly gets overwhelmed. As a result, people usually configure checks to only
+be executed every couple of minutes, or they run into more serious problems.
+
+However, Prometheus takes a fundamentally different approach altogether.
+Instead of executing check scripts, it only collects time series data from a
+set of instrumented targets over the network. For each target, the Prometheus
+server simply fetches the current state of all metrics of that target over HTTP
+and has no other execution overhead that would be pull-related. This brings us
+to the next point:
+
+## It doesn't matter who initiates the connection
+
+For scaling purposes, it doesn't matter who initiates the TCP connection over
+which metrics are then transferred. Either way you do it, the effort for
+establishing a connection is small compared to the metrics payload and other
+required work.
+
+But a push-based approach could use UDP and avoid connection establishment
+altogether, you say! True, but the TCP/HTTP overhead in Prometheus is still
+negligible compared to the other work that the Prometheus server has to do to
+ingest data (especially persisting time series data on disk). To put some
+numbers behind this: a single beefy Prometheus server can easily store millions
+of time series, with a record of 800,000 incoming samples per second (as
+measured with real production metrics data at SoundCloud). This allows you to
+monitor over 10,000 machines from a single Prometheus server. The scaling
+bottleneck here has never been related to pulling metrics, but usually to the
+speed at which the Prometheus server can ingest the data into memory and then
+sustainably persist it on disk/SSD.
```

Comment:
Not sure if you want to mention it, but the real bottleneck is the purging, not so much the writing. Only once purging kicks fully in, we hit the 800,000 limit.
I guess it would confuse people even more. But perhaps hint towards it like "sustainably persist and expire it on disk".